x/bank/simulation/msgs.go

```diff
@@ -34,26 +53,124 @@ func SingleInputSendTx(mapper auth.AccountKeeper) simulation.Operation {
+func createSendMsg(r *rand.Rand, ctx sdk.Context, accs []simulation.Account, mapper auth.AccountKeeper) (fromAcc simulation.Account, action string, msg bank.MsgSend, abort bool) {
+	fromAcc = simulation.RandomAcc(r, accs)
+	toAcc := simulation.RandomAcc(r, accs)
+	// Disallow sending money to yourself
+	for {
+		if !fromAcc.PubKey.Equals(toAcc.PubKey) {
+			break
+		}
+		toAcc = simulation.RandomAcc(r, accs)
+	}
+	toAddr := toAcc.Address
+	initFromCoins := mapper.GetAccount(ctx, fromAcc.Address).SpendableCoins(ctx.BlockHeader().Time)
+
+	if len(initFromCoins) == 0 {
+		return fromAcc, "skipping, no coins at all", msg, true
+	}
+
+	denomIndex := r.Intn(len(initFromCoins))
+	amt, goErr := randPositiveInt(r, initFromCoins[denomIndex].Amount)
+	if goErr != nil {
+		return fromAcc, "skipping bank send due to account having no coins of denomination " + initFromCoins[denomIndex].Denom, msg, true
+	}
+
+	action = fmt.Sprintf("%s is sending %s %s to %s",
+		fromAcc.Address.String(),
+		amt.String(),
+		initFromCoins[denomIndex].Denom,
+		toAddr.String(),
+	)
+
+	coins := sdk.Coins{sdk.NewCoin(initFromCoins[denomIndex].Denom, amt)}
+	msg = bank.NewMsgSend(fromAcc.Address, toAcc.Address, coins)
+	return
+}
+
+// Sends and verifies the transition of a msg send.
+func sendAndVerifyMsgSend(app *baseapp.BaseApp, mapper auth.AccountKeeper, msg bank.MsgSend, ctx sdk.Context, privkeys []crypto.PrivKey, handler sdk.Handler) error {
+	fromAcc := mapper.GetAccount(ctx, msg.FromAddress)
+	AccountNumbers := []uint64{fromAcc.GetAccountNumber()}
+	SequenceNumbers := []uint64{fromAcc.GetSequence()}
+	initialFromAddrCoins := fromAcc.GetCoins()
+
+	toAcc := mapper.GetAccount(ctx, msg.ToAddress)
+	initialToAddrCoins := toAcc.GetCoins()
+
+	if handler != nil {
+		res := handler(ctx, msg)
+		if !res.IsOK() {
+			if res.Code == bank.CodeSendDisabled {
+				return nil
+			}
+			// TODO: Do this in a more 'canonical' way
+			return fmt.Errorf("handling msg failed %v", res)
+		}
+	} else {
+		tx := mock.GenTx([]sdk.Msg{msg},
+			AccountNumbers,
+			SequenceNumbers,
+			privkeys...)
+		res := app.Deliver(tx)
+		if !res.IsOK() {
+			// TODO: Do this in a more 'canonical' way
+			return fmt.Errorf("Deliver failed %v", res)
+		}
+	}
+
+	fromAcc = mapper.GetAccount(ctx, msg.FromAddress)
+	toAcc = mapper.GetAccount(ctx, msg.ToAddress)
+
+	if !initialFromAddrCoins.Minus(msg.Amount).IsEqual(fromAcc.GetCoins()) {
+		return fmt.Errorf("fromAddress %s had an incorrect amount of coins", fromAcc.GetAddress())
+	}
+
+	if !initialToAddrCoins.Plus(msg.Amount).IsEqual(toAcc.GetCoins()) {
+		return fmt.Errorf("toAddress %s had an incorrect amount of coins", toAcc.GetAddress())
+	}
+
+	return nil
+}
+
+// SingleInputSendTx tests and runs a single msg multisend w/ auth, with one input and one output, where both
+// accounts already exist.
+func SingleInputMultiSendTx(mapper auth.AccountKeeper) simulation.Operation {
```

Comment:
What's the diff between this and `SingleInputMultiSendMsg`? Somewhat confusing.

Reply:
`SingleInputMultiSendMsg` takes in a Bank Keeper and Auth Keeper while `SingleInputMultiSendTx` only takes in an Auth Keeper.  I'm not sure as to the purpose of this, but it was already there, so I just kept it.